File: Python solutions/102. Binary Tree Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        # recursive
        # time complexity: o(n)
        # space complexity: o(n
        # levels = [] 
        # if root == None: return levels        
        # def helper(node, level):
        #     if len(levels) == level: levels.append([])
        #     levels[level].append(node.val)
        #     if node.left != None: helper(node.left, level + 1)
        #     if node.right != None: helper(node.right, level + 1)    
        # helper(root, 0)
        # return levels
        
        # iterative using list as a queue
        # levels = []
        # if root is None: return levels
        # q = []
        # q.append(root)
        # while len(q) > 0:
        #     level = []
        #     size = len(q)
        #     while size > 0:
        #         node = q.pop(0)
        #         level.append(node.val)
        #         if node.left != None: q.append(node.left)
        #         if node.right != None: q.append(node.right)
        #         size -= 1
        #     levels.append(level)
        # return levels
        
        # iterative optimization with deque()
        from collections import deque
        levels = []
        if root is None: return levels
        q = deque()
        q.append(root)
        while len(q) > 0:
            level = []
            size = len(q)
            while size > 0:
                node = q.popleft()
                level.append(node.val)
                if node.left is not None: q.append(node.left)
                if node.right is not None: q.append(node.right)
                size -= 1
            levels.append(level)
        return levels
                
        

File: Java solutions/102. Binary Tree Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<List<Integer>> levels = new ArrayList<List<Integer>>();
    
    public void helper(TreeNode node, int level){
        // start the current level
        if (levels.size() == level) levels.add(new ArrayList<Integer>());
        // fulfill the current level
        levels.get(level).add(node.val);
        // process child nodes for the next level
        if (node.left != null) helper(node.left, level + 1);
        if (node.right != null) helper(node.right, level + 1);
    }
    
    public List<List<Integer>> levelOrder(TreeNode root) {
        // recursive
        // time complexity: o(n)
        // space complexity: o(n)
        // if (root == null) return levels;
        // helper(root, 0);
        // return levels;
        
        // iterative
        // time complexity: o(n)
        // space complexity: o(n)
        List<List<Integer>> res = new ArrayList<>();  
        if (root == null) return res;  
        Queue<TreeNode> q = new LinkedList<>();  
        q.offer(root);  
        while (!q.isEmpty()) {  
            List<Integer> level = new ArrayList<>();  
            int size = q.size();  
            while (size > 0){
            // for (int i = 0; i < cnt; i++) {  
                TreeNode node = q.poll();  
                level.add(node.val);  
                if (node.left != null) q.offer(node.left);  
                if (node.right != null) q.offer(node.right);  
                size--;
            }
            res.add(level);   
        }  
        return res;
    }
}
